app/tiebout.hoon

```diff
@@ -0,0 +1,253 @@
+  ::
+      {%our @ @}
+    ?>  ?=(%circle -.piz)
+    =/  nom/name:hall  i.t.wir
+    =/  red/@ud  red.loc.cos.piz
+    [~ this(circles.tie.sta (~(put by circles.tie.sta) nom red))]
+  ==
+::
+::  +reap: recieve acknowledgement for peer
+::
+++  reap
+  |=  [wir=wire err=(unit tang)]
+  ^-  (quip move _this)
+  ::~&  reap+[wir =(~ err)]
+  ?~  err
+    ::  XX send message to users inbox
+    [~ this]
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ?+  i.wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+      %our
+    ?<  ?=(~ t.wir)
+    :_  this
+    ~
+  ==
+::
+::  +quit: receive subscription failed, resubscribe
+::
+++  quit
+  |=  wir=wire
+  ^-  (quip move _this)
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ?+  i.wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+      %our
+    ?<  ?=(~ t.wir)
+    :_  this
+    =/  doc/dock  [our.bol %hall]
+    [ost.bol %peer /our/[i.t.wir] doc /circle/[i.t.wir]/config/grams]~
+  ==
+
+::
+::  +diff-hall-rumor: receive message or a read event from a hall circle
+::
+++  diff-hall-rumor
+  |=  [wir=wire rum=rumor:hall]
+  ^-  (quip move _this)
+  ::~&  rumor+[wir rum]
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ?+  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ::
+  ::  %our
+  ::
+      {%our @ @}
+    ?>  ?=(%circle -.rum)
+    =/  nom/name:hall  i.t.wir
+    ?+  -.rum.rum
+      [~ this]
+    ::
+    ::  %gram: send notification if envelope is lower than read number
+    ::
+      %gram
+    =/  red  (~(get by circles.tie.sta) nom)
+    ?~  red
+      (mean [leaf+"invalid circle for diff: {(spud wir)}"]~)
+    ?:  (gth num.nev.rum.rum u.red)
+      :_  this(circles.tie.sta (~(put by circles.tie.sta) nom u.red))
+      (message-to-notification u.red nev.rum.rum)
+    :_  this
+    (message-to-notification u.red nev.rum.rum)
+    ::
+    ::   %config: set our read number
+    ::
+      %config
+    ?+  -.dif.rum.rum
+      [~ this]
+          %read
+        =/  red/@ud  red.dif.rum.rum
+        [~ this(circles.tie.sta (~(put by circles.tie.sta) nom red))]
+      ==
+    ==
+  ==
+::
+::  generate notification move from hall message
+::
+++  message-to-notification
+  |=  [red=@ud env=envelope:hall]
+  ^-  (list move)
+  ?:  =(aut.gam.env our.bol)
+    ~
+  =/  pay  %-  my  :~
+    alert+s+'New message from {(cite:title aut.gam.env)}'
+  ==
+  =/  not/notification  :+
+    token.tie.sta
+    'com.tlon.urbit-client'
+    pay
+  ?:  (lte num.env red)
+    ~
+  =/  doc/dock  [king.tie.sta dap.bol]
+  [ost.bol %poke /ask-king doc %tiebout-action [%notify not]]~
+
+::
+::  +create-apns-request: create hiss with payload for APNs
+::
+++  create-apns-request
+  |=  not=notification
+  ^-  hiss
+  =/  furl=@t  (crip (weld (trip baseurl.tie.sta) (trip token.not)))
+  =/  url=purl  (need (de-purl:html furl))
+  =/  jon=json  :-  %o  %-  my  :~
+    aps+o+payload.not
+  ==
```

Comment:
[o, my payload](https://i0.kym-cdn.com/entries/icons/original/000/023/390/O.png)

Same map building advice from above applies here (and below). But also, there's stdlib functions dedicated to construction json. Look at `++en-json`.

(Also chaining tall-form together on the same line like this is generally bad style.)

Reply:
Will fix tall-form chaining.